### PR TITLE
BREAKING CHANGE: split article head

### PR DESCRIPTION
**Pull request by jimhunty**
https://nidigitalsolutions.jira.com/browse/REPLAT-3978

Essentially this takes the header our of `article` and puts it into `article-main-standard`. `article` is then cleaned up and accepts a header as a prop so we can (on native) conform to one flatlist. There has been an effort made to move items that might repeat in other templates into other packages.

****Additional bits that were done****
* Ad config in fixtures centralised
* Fixing webpack errors
* Web articles do not need props like `onAuthorPress`
* Article Topics currently exists in both places. - This is fine due to the different style implementations
* Headline knob added.
* Aspect ratio container centralised
* Article Loading now removed.
* Article Topics show top and bottom on destop
* Mobile and tablet web image now on top
* getLeadAsset needs centralised
* Design decisions made on `onViewableItems` and `receiveChildList` to be passed from the parent

****To Do Later****
* Button designs on Android are off - need a fix
* Move knobs to the right on storybook for `article` and `article-main-standard`. (Separate story) https://github.com/storybook-eol/storybook-addon-knobs/issues/29
* Fix React strict mode warnings/errors.
* Related Articles should manage its own empty state
* Jest serialiser for ForwardRef.
* Label is bumping into headline on card (Android)
* Be a good idea to have test tools for all templates as there will be a large amount of repetition.
* Showcase Helper needs centralising? Needs investigating
* Null lead asset throws an error? Needs investigating
* Manage fixtures better.
* Get Headline unit test required
* Lead Asset Crop Config move back? Aron has more details

### CODE REVIEW

File: packages/article-main-standard/__tests__/mocks.native.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { mockNativeModules } from "@times-components/test-utils";
```
Comment (nehasri89) [inline packages/article-main-standard/__tests__/mocks.native.js]: can see same stuff in this and .web file, might want to DRY it up

File: packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap [hunk 2]
```diff
@@ -0,0 +1,157 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`1. a full article with an image as the lead asset 1`] = `
+<Article
+  adConfig={
+    Object {
+      "adUnit": "mockAdUnit",
+      "bidderSlots": Array [
```
Comment (nehasri89) [inline packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap]: don't see an image here, also is there a way to get rid of the huge adConfig object ?

File: packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        "path": "/news",
+      },
+    }
+  }
+/>
+`;
+
+exports[`2. an article with a video as the lead asset 1`] = `
+<Article
+  adConfig={
```
Comment (nehasri89) [inline packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap]: no video? same as above

File: packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  }
+/>
+`;
+
+exports[`3. an article with no lead asset 1`] = `
+<Article
+  adConfig={
+    Object {
+      "adUnit": "mockAdUnit",
+      "bidderSlots": Array [
```
Comment (nehasri89) [inline packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap]: same as above

File: packages/article-main-standard/__tests__/mocks.native.js (hunk 1 above)
Comment (jimhunty) [reply packages/article-main-standard/__tests__/mocks.native.js]: I would suggest looking at this as a seperate PR, this was a copy over from `article` and needs a good look at and some love.

File: packages/article-topics/src/article-topic-prop-types.js [hunk 5]
```diff
@@ -1,9 +1,11 @@
 import PropTypes from "prop-types";
 
-const propTypes = {
+const topicPropTypes = {
+  fontSize: PropTypes.number,
+  lineHeight: PropTypes.number,
   name: PropTypes.string.isRequired,
   onPress: PropTypes.func.isRequired,
-  slug: PropTypes.string.isRequired
+  slug: PropTypes.string
```
[the remark this answers is not in the document]
Comment (jimhunty) [reply packages/article-topics/src/article-topic-prop-types.js]: I've put this back. Seems like a mistake.

File: packages/article-main-standard/__tests__/android/__snapshots__/article-with-style.android.test.js.snap [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        </Text>
+      </View>
+      <Button />
+    </View>
+  </View>
+</RCTScrollView>
+`;
+
+exports[`2. a full article with an image as the lead asset 1`] = `<Article />`;
+
```
Comment (nehasri89) [inline packages/article-main-standard/__tests__/android/__snapshots__/article-with-style.android.test.js.snap]: same as above

File: packages/article-main-standard/__tests__/android/__snapshots__/article.android.test.js.snap [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+      "biddersConfig": Object {
+        "bidders": Object {
+          "amazon": Object {
+            "accountId": "3360",
+          },
+          "appnexus": Object {
+            "placementId": "5823281",
+          },
+          "criteo": Object {
+            "zoneMap": Object {
```
Comment (nehasri89) [inline packages/article-main-standard/__tests__/android/__snapshots__/article.android.test.js.snap]: no lead asset.

File: packages/article-main-standard/__tests__/ios/__snapshots__/article-lead-asset.ios.test.js.snap [hunk 8]
```diff
@@ -0,0 +1,157 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`1. a full article with an image as the lead asset 1`] = `
```
Comment (nehasri89) [inline packages/article-main-standard/__tests__/ios/__snapshots__/article-lead-asset.ios.test.js.snap]: needs fixing

File: packages/article-main-standard/__tests__/mocks.native.js (hunk 1 above)
Comment (nehasri89) [reply packages/article-main-standard/__tests__/mocks.native.js]: cool

File: packages/article-main-standard/src/article-main-standard.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    if (error) {
+      return <ArticleError refetch={refetch} />;
+    }
+
+    if (isLoading) {
+      return null;
+    }
+
+    return (
+      <AdComposer adConfig={this.props.adConfig}>
```
Comment (nehasri89) [inline packages/article-main-standard/src/article-main-standard.js]: destructure more of `this.props` on line 99?

File: packages/article-main-standard/src/article-page-prop-types.js [hunk 10]
```diff
@@ -16,17 +18,13 @@ const articlePagePropTypes = {
     })
   }),
   isLoading: PropTypes.bool,
-  onRelatedArticlePress: PropTypes.func.isRequired,
```
Comment (nehasri89) [inline packages/article-main-standard/src/article-page-prop-types.js]: think we still need them?

File: packages/article-main-standard/src/article-main-standard.js (hunk 9 above)
Comment (jimhunty) [reply packages/article-main-standard/src/article-main-standard.js]: done.

File: packages/article-main-standard/src/article-page-prop-types.js (hunk 10 above)
Comment (jimhunty) [reply packages/article-main-standard/src/article-page-prop-types.js]: Only on native. Implemented this better now.

File: packages/article-main-standard/__tests__/ios/__snapshots__/article-lead-asset.ios.test.js.snap (hunk 8 above)
Comment (jimhunty) [reply packages/article-main-standard/__tests__/ios/__snapshots__/article-lead-asset.ios.test.js.snap]: Done

File: packages/article-main-standard/__tests__/android/__snapshots__/article.android.test.js.snap [hunk 11]
```diff
@@ -0,0 +1,394 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
[the remark this answers is not in the document]
Comment (jimhunty) [reply packages/article-main-standard/__tests__/android/__snapshots__/article.android.test.js.snap]: done

File: packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap (hunk 2 above)
Comment (jimhunty) [reply packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap]: done

File: packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap (hunk 3 above)
Comment (jimhunty) [reply packages/article-main-standard/__tests__/android/__snapshots__/article-lead-asset.android.test.js.snap]: done

File: packages/article-main-standard/__tests__/ios/__snapshots__/article.ios.test.js.snap [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+  windowSize={21}
+>
+  <View>
+    <View>
+      <View />
+      <View>
+        <Text
+          selectable={true}
+        >
+          Some Short Headline
```
Comment (craigbilner) [inline packages/article-main-standard/__tests__/ios/__snapshots__/article.ios.test.js.snap]: these types of test should ideally focus on that which is under test

File: packages/article-main-standard/__tests__/shared-with-style.native.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+      print,
+      minimalNativeTransform,
+      minimaliseTransform((value, key) => key !== "style"),
+      flattenStyleTransform
+    )
+  );
+
+  it("full article with style", () => {
+    const article = articleFixture({
+      ...testFixture,
```
Comment (craigbilner) [inline packages/article-main-standard/__tests__/shared-with-style.native.js]: should move to schema mocks

File: packages/article-main-standard/__tests__/web/__snapshots__/semantic.web.test.js.snap [hunk 14]
```diff
@@ -18,71 +18,69 @@ exports[`1. a full article with an image as the lead asset 1`] = `
...
-        Diamond Icon
-      </title>
-      <g>
-        <path
-          d="90928a4d6e9b702222e7187e68c282b7"
-        />
-      </g>
-    </svg>
-    new
+  SOME LABEL
```
Comment (craigbilner) [inline packages/article-main-standard/__tests__/web/__snapshots__/semantic.web.test.js.snap]: is this correct?

File: packages/article-main-standard/src/article-main-standard.web.js [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+            />
+            <ArticleTopics topics={topics} />
+          </MetaContainer>
+          <LeadAssetContainer>
+            <LeadAssetComponent
+              {...leadAssetProps}
+              width={this.state.articleWidth}
+            />
+          </LeadAssetContainer>
+          <BodyContainer>
```
Comment (craigbilner) [inline packages/article-main-standard/src/article-main-standard.web.js]: I was expecting the inversion of control to be with `Article` here and the templates  to basically just do the head?

File: packages/article-main-standard/README.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
+# Article Main Standard (Template)
+
+This package is the default template for Articles. Its main responsibility is to manage its own header and uses the `article` package to render its body content, essentially creating a composed dumb component of a page.
+
+It is also responsible for tracking and ads.
```
[the remark this answers is not in the document]
Comment (jimhunty) [reply packages/article-main-standard/README.md]: This is a note for other templates to make sure tracking and ads are included as they are easily forgotten. If the responsibility for that doesn't lie here, then where?

File: packages/article-main-standard/__tests__/ios/__snapshots__/article.ios.test.js.snap (hunk 12 above)
Comment (jimhunty) [reply packages/article-main-standard/__tests__/ios/__snapshots__/article.ios.test.js.snap]: This was ported across from Article. There are plans to improve this.

File: packages/article-main-standard/__tests__/shared-with-style.native.js (hunk 13 above)
Comment (jimhunty) [reply packages/article-main-standard/__tests__/shared-with-style.native.js]: Fixtures is on the to-do list.

File: packages/article-main-standard/README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (aronshamash) [reply packages/article-main-standard/README.md]: Can we remove this line now ?

File: packages/article/src/article.web.js [hunk 17]
```diff
@@ -1,212 +1,93 @@
...
+                section={section}
+              />
+            </BodyContainer>
+            <ArticleTopics topics={topics} />
+            <aside id="related-articles" ref={node => registerNode(node)}>
+              {displayRelatedArticles({
+                isVisible: !!observed.get("related-articles")
+              })}
+            </aside>
+            <Ad contextUrl={url} section={section} slotName="pixel" />
```
Comment (aronshamash) [inline packages/article/src/article.web.js]: minor indentation issue but better having Ads in `article` now 👍

File: packages/article-main-standard/src/article-main-standard.web.js (hunk 15 above)
Comment (jimhunty) [reply packages/article-main-standard/src/article-main-standard.web.js]: Changed!

**Pull request merged**